api/src/main/java/org/apache/iceberg/Table.java

```diff
@@ -44,6 +44,15 @@
    */
   TableScan newScan();
 
+
+  /**
+   * @param fromSnapshotId - the last snapshot id read by the user, exclusive
+   * @param toSnapshotId - read incremental data upto this snapshot id
+   * @return a table scan which can read incremental data from {@param fromSnapshotId}
+   * exclusive and up to {@toSnapshotId} inclusive
+   */
+  TableScan newIncrementalScan(long fromSnapshotId, long toSnapshotId);
```

Comment:
TODO: What about when the user reads incremental scan for the first time [bootstrap] what should the `fromSnapshotId` be , or even `toSnapshotId` ?

Reply:
For new tables, from will be the first snapshot ID. For existing tables, where the oldest snapshot has already expired, users will need to choose a starting snapshot. Usually, that would be after running a batch process to handle existing table data before the incremental process starts. So it would be the snapshot on which the batch process ran.

Reply:
I think it is essential that we can start a stream data from a large table and have multiple batches for this. @rdblue, can you elaborate on the batch process you mentioned?

I've also summarized my thoughts on requirements for Structured Streaming sources in [this](https://github.com/apache/incubator-iceberg/issues/179#issuecomment-554666093) comment. Let me know if that makes sense to you, @rdsr @rdblue.

Reply:
Splitting a snapshot into multiple batches will require some thinking. Maybe, this can be done by streaming sources. For example, our implementation of `Offset` in Spark can store additional information. Storing consumed files in offset JSON files isn't an option but maybe we can store a pointer to a list of consumed manifests or something (which can be an Avro or a Parquet file, for example).

Reply:
That sounds reasonable to me, but I think it should also be possible to start consuming from a specific snapshot, since processing the rest of the table seems needlessly expensive in a lot of cases. We move jobs from batch to streaming and don't want to rewrite the history.

Reply:
Absolutely agree, consuming from a given snapshot covers most of use cases. At the same time, there are scenarios when Structured Streaming in Spark is used, for example, beyond streaming. I just want to capture those use cases as well.

Reply:
When using incremental for the first time [bootstrap] I think  `fromSnapshotId` should be null since the contract is  an incremental scan gives all the appends between `fromSnapshotId` **exclusive** to `toSnapshotId`  .  When `fromSnapshotId` is null we will return all the appends till `toSnapshotId`. In this way a user will not miss any data.

thoughts?

Reply:
How about using `Table#newScan` for the first time, and then call `TableScan#snapshot` to know till what snapshot the first time read?